pkg/controller/controller.go

```diff
@@ -430,55 +463,106 @@ func (c *Controller) getDesiredWorkers(
 	idleWorkers int32,
 	minWorkers int32,
 	maxWorkers int32,
-	minAvailableWorkers int32) int32 {
+	maxDisruption *string) int32 {
 
 	// overwrite the minimum workers needed based on
 	// messagesSentPerMinute and secondsToProcessOneJob
 	// this feature is disabled if secondsToProcessOneJob is not set or is 0.0
-	minWorkers = c.getMinWorkers(messagesSentPerMinute, minWorkers, secondsToProcessOneJob)
+	minWorkers = getMinWorkers(
+		messagesSentPerMinute,
+		minWorkers,
+		secondsToProcessOneJob,
+	)
+
+	// gets the maximum number of workers that can be scaled down in a
+	// single scale down activity.
+	maxDisruptableWorkers := getMaxDisruptableWorkers(
+		maxDisruption, currentWorkers,
+	)
 
 	tolerance := 0.1
 	usageRatio := float64(queueMessages) / float64(targetMessagesPerWorker)
 	if currentWorkers == 0 {
 		desiredWorkers := int32(math.Ceil(usageRatio))
-		return convertDesiredReplicasWithRules(desiredWorkers, minWorkers, maxWorkers)
+		return convertDesiredReplicasWithRules(
+			currentWorkers,
+			desiredWorkers,
+			minWorkers,
+			maxWorkers,
+			maxDisruptableWorkers,
+		)
 	}
 
 	if queueMessages > 0 {
 		// return the current replicas if the change would be too small
-		if math.Abs(1.0-usageRatio) <= tolerance {
-			return currentWorkers
-		}
-
-		if queueMessages < targetMessagesPerWorker {
-			return currentWorkers
+		if (math.Abs(1.0-usageRatio) <= tolerance) || (queueMessages < targetMessagesPerWorker) {
+			// desired is same as current in this scenario
+			return convertDesiredReplicasWithRules(
+				currentWorkers,
+				currentWorkers,
+				minWorkers,
+				maxWorkers,
+				maxDisruptableWorkers,
+			)
 		}
 
 		desiredWorkers := int32(math.Ceil(usageRatio * float64(currentWorkers)))
-		if desiredWorkers < minAvailableWorkers {
-			desiredWorkers = minAvailableWorkers
-		}
-
-		return convertDesiredReplicasWithRules(desiredWorkers, minWorkers, maxWorkers)
+		return convertDesiredReplicasWithRules(
+			currentWorkers,
+			desiredWorkers,
+			minWorkers,
+			maxWorkers,
+			maxDisruptableWorkers,
+		)
 	} else if messagesSentPerMinute > 0 && secondsToProcessOneJob > 0.0 {
 		// this is the case in which there is no backlog visible.
 		// (mostly because the workers picks up jobs very quickly)
 		// But the queue has throughput, so we return the minWorkers.
 		// Note: minWorkers is updated based on
 		// messagesSentPerMinute and secondsToProcessOneJob
-		return minWorkers
+		// desried is the minReplicas in this scenario
+		return convertDesiredReplicasWithRules(
+			currentWorkers,
+			minWorkers,
+			minWorkers,
+			maxWorkers,
+			maxDisruptableWorkers,
+		)
 	}
 
 	// Attempt for massive scale down
 	if idleWorkers > 0 {
 		desiredWorkers := currentWorkers - idleWorkers
-		return convertDesiredReplicasWithRules(desiredWorkers, minWorkers, maxWorkers)
+		return convertDesiredReplicasWithRules(
+			currentWorkers,
+			desiredWorkers,
+			minWorkers,
+			maxWorkers,
+			maxDisruptableWorkers,
```

Comment:
should be hardcoded to 100% so that complete scale down based on all workers being idle can happen.